Scripts/Runtime/Collections/DeferredNativeArray.cs

```diff
@@ -0,0 +1,333 @@
+
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+            NativeArrayUnsafeUtility.SetAtomicSafetyHandle(ref array, m_Safety);
+#endif
+            return array;
+        }
+
+        /// <summary>
+        /// Returns a <see cref="NativeArray{T}"/> for use in a job.
+        /// Initially this <see cref="NativeArray{T}"/> will not have anything in it but later on after
+        /// <see cref="DeferredCreate"/> is called, the <see cref="NativeArray{T}"/> instance returned will point to
+        /// the same memory. By scheduling your jobs, you can have a job populate the <see cref="DeferredNativeArray{T}"/>
+        /// and a subsequent job read from this <see cref="NativeArray{T}"/> before you know the full length.
+        /// </summary>
+        /// <returns>A <see cref="NativeArray{T}"/> instance that will be populated in the future.</returns>
+        public unsafe NativeArray<T> AsDeferredJobArray()
+        {
+            AssertForAsDeferredJobArray();
+
+            //This whole function taken from NativeList.AsDeferredJobArray
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+            AtomicSafetyHandle.CheckExistsAndThrow(m_Safety);
+#endif
+            byte* buffer = (byte*)m_BufferInfo;
+            // Unity uses this as an indicator to the internal Job Scheduling code that it needs to defer scheduling until
+            // the array length is actually known. 
+            buffer += 1;
+            NativeArray<T> array = NativeArrayUnsafeUtility.ConvertExistingDataToNativeArray<T>(buffer, 0, Allocator.Invalid);
+
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+            NativeArrayUnsafeUtility.SetAtomicSafetyHandle(ref array, m_Safety);
+#endif
+
+            return array;
+        }
+
+        //*************************************************************************************************************
+        // JOBS
+        //*************************************************************************************************************
+
+        [BurstCompile]
+        private readonly unsafe struct DisposeJob : IJob
+        {
+            [NativeDisableUnsafePtrRestriction] private readonly DeferredNativeArrayBufferInfo* m_BufferInfo;
+            private readonly Allocator m_Allocator;
+
+            public DisposeJob(DeferredNativeArrayBufferInfo* bufferInfo, Allocator allocator)
+            {
+                m_BufferInfo = bufferInfo;
+                m_Allocator = allocator;
+            }
+
+            public void Execute()
+            {
+                //This dispose job just handles freeing the memory, the other aspects of the collection were already
+                //when this job was scheduled because it requires main thread access
+                UnsafeUtility.Free(m_BufferInfo->Buffer, m_Allocator);
+                UnsafeUtility.Free(m_BufferInfo, m_Allocator);
+            }
+        }
+    }
+
+    [BurstCompatible]
+    public static unsafe class DeferredNativeArrayUnsafeUtility
+    {
```

Comment:
They are public because we need them for scheduling jobs. 

`BufferInfo` is internal, but we're never returning the struct to operate on, just a `void*` so that the scheduler knows which memory to look at. 

The scheduler is dumb. It just says 
"ok so at this address, there should be void* so let's move forward and then next thing should be the length. Cool I'll check this later on"